pkg/resource/user/delta_util.go

```diff
@@ -23,12 +23,10 @@ func filterDelta(
 ) {
 	// the returned AccessString can be different than the specified one; as long as the last requested AccessString
 	// matches the currently desired one, remove this difference from the delta
-	//TODO: revert this call to Spec.AccessString once we have a new implementation of it
-	if delta.DifferentAt("AccessString") {
+	if delta.DifferentAt("Spec.AccessString") {
 		if *desired.ko.Spec.AccessString == *desired.ko.Status.LastRequestedAccessString {
 
-			//TODO: revert the call to Spec.AccessString once removeFromDelta implementation changes
-			removeFromDelta(delta, "AccessString")
+			removeFromDelta(delta, "Spec.AccessString")
```

Comment:
Why do you have multiple "AccessString" fields in the `Status` struct?

https://github.com/aws-controllers-k8s/elasticache-controller/blob/e0d2cb95702866ec24b7147a9b52d2bea8f5ba06/generator.yaml#L122-L131

Reply:
The access string provided by the ElastiCache API differs from the specified access string (there's some behind-the-scenes expansion). 

Because there's no straightforward way to translate between the two representations, adding a state-keeping field `Status.LastRequestedAccessString` allows us to determine when the desired access string has changed. This field is only updated upon a successful create or update (see the `set_output` methods).

Since the generated `delta` code will always detect a difference between `desired.Spec.AccessString` and `latest.Spec.AccessString` there's a hook after that to then remove the difference if `desired.Spec.AccessString` is still equal to `desired.Status.LastRequestAccessString`. See `filterDelta()` in `pkg/resource/user/delta_util.go`.

`Status.ExpandedAccessString` is purely informational and was added because that's what the user would see if they were directly using the ElastiCache API.

Reply:
@echen-98 why aren't you simply updating `Spec.AccessString` to the actual expanded value upon return from CreateUser and just get rid of the two other fields in the Status?

Reply:
@jaypipes do you mean updating `desired.Spec.AccessString` or `latest.Spec.AccessString`? We're already doing the latter, and unless I'm missing something big, I thought the former wasn't possible?

Reply:
i.e. I thought `patchResource()` only patched `Status` because `desired.Spec` should not be modified

Reply:
> @jaypipes do you mean updating `desired.Spec.AccessString` or `latest.Spec.AccessString`? We're already doing the latter, and unless I'm missing something big, I thought the former wasn't possible?

`latest.Spec.AccessString`

If you're already doing that, why do you need multiple copies of that value in `Status.LastRequestedAccessString` and `Status.ExpandedAccessString`?

Reply:
@jaypipes

`desired.Spec.AccessString`: `on ~app::* -@all +@read`
`latest.Spec.AccessString`: `on ~app::* -@all +@read +@hash +@bitmap +@geo -setbit -bitfield -hset -hsetnx -hmset -hincrby -hincrbyfloat -hdel -bitop -geoadd -georadius -georadiusbymember`

There is always a `Spec` difference in `AccessString`. Without any custom code, this path gets added to the `delta`, and an Update is always invoked. So the default behavior of the generated code is to make Update calls in an infinite loop.

Even if we introduce custom update code, we have no way of telling whether we should update or not (since as I said above, we can't translate between the two representations). So this leaves us no better than the generated code.

The only way to tell if the desired state has changed is to add a state-keeping field such as `Status.LastRequestedAccessString`. Although this path will always be added to the delta, this allows us to remove it from the delta if there has been no change, and therefore not invoke an Update at all. If there is a difference, then the path is left in the delta and the resulting Update call is actually meaningful.

As for `Status.ExpandedAccessString`, `latest.Spec` is not visible to the user, so the only way to make it visible is to make it a `Status` field. This is actually a larger issue with other resources as well.

Let me know if my explanation is lacking anywhere and I can try to elaborate further.

Reply:
> @jaypipes
> 
> `desired.Spec.AccessString`: `on ~app::* -@all +@read`
> `latest.Spec.AccessString`: `on ~app::* -@all +@read +@hash +@bitmap +@geo -setbit -bitfield -hset -hsetnx -hmset -hincrby -hincrbyfloat -hdel -bitop -geoadd -georadius -georadiusbymember`
> 
> There is always a `Spec` difference in `AccessString`. Without any custom code, this path gets added to the `delta`, and an Update is always invoked. So the default behavior of the generated code is to make (mostly useless) Update calls in an infinite loop.

If we were saving `Spec.AccessString` as the value returned by the server after a CreateUser or UpdateUser call, there would be no infinite loop.

Reply:
So in the Create case, it was my impression that in the reconciler, after a resource is found to not exist and a Create is called, then a ReadOne will be called next. If this ReadOne populates `latest.Spec.AccessString` with the expanded access string, there will be a `Spec` difference (since `desired.Spec` has the user-requested access string). And as long as there is a `Spec` difference, then `Update` will be invoked.

What I'm trying to say is it doesn't matter whether we save the server-returned value in `latest.Spec.AccessString` in a Create or Update call (especially since `latest.Spec` is thrown away upon patching in these two cases). The infinite loop depends solely upon what happens in `ReadOne`, and `ReadOne` will always populate `latest.Spec.AccessString` with a value different from `desired.Spec.AccessString`.

Reply:
Sorry, I should add that there may not be an infinite loop (that was the initial behavior when I submitted my first iteration of the User code before the split repos, and there have been several runtime changes since then). However, this would still cause an unnecessary Update call after the creation of the resource, because the ReadOne right after the Create would pick up the `Spec` difference. 

Regardless, the point is that we need to store the last requested access string. Otherwise we won't know when the desired access string has changed. If the user updates another field other than the access string, we want to know whether or not to add the access string to the Update payload (I don't think it's best practice to add a field unless it actually has changed).

This is the approach that we decided on before the split repos, and we decided that we should keep it once the User resource was brought back into scope. It works, and if it's sub-optimal, that just reflects that we sometimes have to do sub-optimal things when dealing with the ElastiCache API.

The real point of this discussion is the more general question of: how do you handle the case where an API returns a value different from the one requested? I think that's a very important question to answer, but I also think that this PR is not the place to have this discussion. The point of this PR was to finally have succeeding Prow runs, and to continue to block it until we have a clear answer to that question doesn't make sense to me.

Also, even if we were to settle on a different solution, that would be in a follow-up PR anyway.

https://github.com/echen-98/ack-community/commit/f0590646edc94fff9e8d58dd0ecca68b86b4f361
https://github.com/aws-controllers-k8s/elasticache-controller/pull/32

Reply:
I agree with @echen-98
We shall let this PR go through and have different issues to track discussion on Spec, Status population and Delta computation and resource patching related aspects for special fields (readonly, server modifiable, not specified in spec but server provided defaults etc.).

Reply:
@jaypipes please suggest if its fine to merge this PR in its current state.